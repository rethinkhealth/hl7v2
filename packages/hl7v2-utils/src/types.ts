/**
 * Diagnostic definition.
 *
 * A diagnostic is a message about a problem in the HL7v2 message. It is used to report issues to the user.
 *
 * Diagnostics are defined per-package with type, namespace, and code that compose the `ruleId`.
 *
 * Diagnostics are used in conjunction with the `report()` function to report issues to the user.
 */
export type Diagnostic = Readonly<{
  /**
   * Tool/plugin category
   */
  type: "validator" | "lint" | "annotator" | "transformer" | "parser" | string;

  /**
   * Domain/concern (order, field, conformance, segment, group, datatype, etc.)
   */
  namespace: string;

  /**
   * Specific issue code (transition, required, acceptance, mismatch, etc.)
   */
  code: string;

  /**
   * Full description of the rule
   */
  description?: string;

  /**
   * Default severity
   *
   * @defaultValue 'info'
   */
  severity?: "error" | "warning" | "info" | null | undefined;

  /**
   * Message formatter function.
   * Takes context and returns the formatted message.
   * The function signature is self-documenting: it shows what context is required.
   */
  message: (context: Record<string, unknown>) => string;

  /**
   * Optional URL to documentation
   */
  helpUrl?: string;
}>;
