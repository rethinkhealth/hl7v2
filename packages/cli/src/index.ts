#!/usr/bin/env node
/** biome-ignore-all lint/suspicious/noConsole: This is a CLI */
import { readFileSync } from 'node:fs';
import { createInterface } from 'node:readline';

import { type HL7v2Node, parseHL7 } from '@rethinkhealth/hl7v2';
import { Command } from 'commander';

const program = new Command();

program.name('hl7').description('HL7v2 CLI Tool').version('0.1.0');

program
  .command('parse')
  .description('Parse an HL7v2 message')
  .argument('[file]', 'Path to HL7 file')
  .action(async (file) => {
    let message: string;

    if (file) {
      message = readFileSync(file, 'utf8');
    } else {
      // Interactive mode
      const rl = createInterface({
        input: process.stdin,
        output: process.stdout,
      });
      message = await new Promise<string>((resolve) =>
        rl.question('Paste HL7 message:\n', resolve)
      );
      rl.close();
    }

    const dom = parseHL7(message);

    // Show parsed tree
    console.log('\nParsed HL7 DOM:\n');
    console.log(JSON.stringify(dom, null, 2));

    // Basic stats
    const segments = dom.children?.length || 0;
    console.log('\nStats:');
    console.log(`- Segments: ${segments}`);
    console.log(`- Characters: ${message.length}`);
    console.log(`- Message Type: ${getMessageType(dom)}`);
  });

program.parse(process.argv);

// Extract MSH-9 for message type
function getMessageType(dom: HL7v2Node): string {
  const msh = dom.children?.find((n: HL7v2Node) => n.name === 'MSH');
  return msh?.children?.[8]?.value || 'Unknown';
}
