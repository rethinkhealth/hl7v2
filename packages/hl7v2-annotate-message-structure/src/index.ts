import type { Root, RootData } from "@rethinkhealth/hl7v2-ast";
import type { MessageInfo } from "@rethinkhealth/hl7v2-util-message-info";
import type { Plugin } from "unified";

/**
 * Unified plugin to infer HL7v2 message structure when missing.
 *
 * This plugin infers the message structure (MSH-9.3) from message code (MSH-9.1)
 * and trigger event (MSH-9.2) when the structure field is not present in the source.
 *
 * The inferred structure follows the convention: `{messageCode}_{triggerEvent}`
 *
 * @example
 * ```typescript
 * import { unified } from 'unified';
 * import { hl7v2Parser } from '@rethinkhealth/hl7v2-parser';
 * import { hl7v2AnnotateMessage } from '@rethinkhealth/hl7v2-annotate-message';
 * import { hl7v2AnnotateMessageStructure } from '@rethinkhealth/hl7v2-annotate-message-structure';
 *
 * const processor = unified()
 *   .use(hl7v2Parser)
 *   .use(hl7v2AnnotateMessage)
 *   .use(hl7v2AnnotateMessageStructure);
 *
 * const tree = processor.parse(message);
 * await processor.run(tree);
 *
 * console.log(tree.data.messageInfo);
 * // { version: "2.5", messageCode: "ADT", triggerEvent: "A01", messageStructure: "ADT_A01" }
 * ```
 */
export const hl7v2AnnotateMessageStructure: Plugin<[], Root, Root> = () => {
  return (tree: Root) => {
    // Ensure data object exists
    if (!tree.data) {
      return tree;
    }

    // Get messageInfo from tree.data (cast to include messageInfo)
    const messageInfo = (tree.data as RootData & { messageInfo?: MessageInfo })
      .messageInfo;

    // If messageInfo doesn't exist, nothing to do
    if (!messageInfo) {
      return tree;
    }

    // If messageStructure is already set, don't override it
    if (messageInfo.messageStructure) {
      return tree;
    }

    // If we have both messageCode and triggerEvent, infer the structure
    if (messageInfo.messageCode && messageInfo.triggerEvent) {
      messageInfo.messageStructure = `${messageInfo.messageCode}_${messageInfo.triggerEvent}`;
    }

    return tree;
  };
};

export default hl7v2AnnotateMessageStructure;
