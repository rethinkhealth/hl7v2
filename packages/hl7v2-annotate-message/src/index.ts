import type { Root } from "@rethinkhealth/hl7v2-ast";
import type { MessageInfo } from "@rethinkhealth/hl7v2-util-message-info";
import { getMessageInfo } from "@rethinkhealth/hl7v2-util-message-info";
import type { Plugin } from "unified";

/**
 * Augment the RootData type to include messageInfo
 */
declare module "@rethinkhealth/hl7v2-ast" {
  // biome-ignore lint/nursery/useConsistentTypeDefinitions: module augmentation
  export interface RootData {
    /**
     * Message metadata extracted from MSH segment
     * Includes version (MSH-12), message code (MSH-9.1),
     * trigger event (MSH-9.2), and message structure (MSH-9.3)
     */
    messageInfo?: MessageInfo;
  }
}

/**
 * Unified plugin to annotate HL7v2 AST with message metadata.
 *
 * This plugin extracts message metadata from the MSH segment and stores it
 * in `Root.data.messageInfo` for downstream plugins to access.
 *
 * The metadata includes:
 * - `version`: MSH-12 (e.g., "2.5")
 * - `messageCode`: MSH-9.1 (e.g., "ADT")
 * - `triggerEvent`: MSH-9.2 (e.g., "A01")
 * - `messageStructure`: MSH-9.3 (e.g., "ADT_A01")
 *
 * @example
 * ```typescript
 * import { unified } from 'unified';
 * import { hl7v2Parser } from '@rethinkhealth/hl7v2-parser';
 * import { hl7v2AnnotateMessage } from '@rethinkhealth/hl7v2-annotate-message';
 *
 * const processor = unified()
 *   .use(hl7v2Parser)
 *   .use(hl7v2AnnotateMessage);
 *
 * const tree = processor.parse(message);
 * await processor.run(tree);
 *
 * console.log(tree.data.messageInfo);
 * // { version: "2.5", messageCode: "ADT", triggerEvent: "A01", messageStructure: "ADT_A01" }
 * ```
 */
export const hl7v2AnnotateMessage: Plugin<[], Root, Root> = () => {
  return (tree: Root) => {
    const messageInfo = getMessageInfo(tree);

    // Initialize data object if it doesn't exist
    if (!tree.data) {
      tree.data = {};
    }

    // Store message info in Root.data
    tree.data.messageInfo = messageInfo;

    return tree;
  };
};

export default hl7v2AnnotateMessage;
