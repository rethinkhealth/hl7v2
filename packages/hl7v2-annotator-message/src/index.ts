/** biome-ignore-all lint/performance/noBarrelFile: fine */
export * from './plugin';
export type * from './types';
