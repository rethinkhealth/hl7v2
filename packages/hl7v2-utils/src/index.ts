/** biome-ignore-all lint/performance/noBarrelFile: fine */
export { report } from "./report";
export type { Diagnostic } from "./types";
export { DEFAULT_DELIMITERS, isEmptyNode } from "./utils";
