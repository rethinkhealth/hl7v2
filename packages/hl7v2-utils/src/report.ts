import type { Node } from "@rethinkhealth/hl7v2-ast";
import type { VFile } from "vfile";
import type { Diagnostic } from "./types";

export type ReportOptions = {
  node?: Node;
  context?: Record<string, unknown>;
  ancestors?: Node[];
};

/**
 * Report a diagnostic to a vfile.
 *
 * Calls the diagnostic's message function with context and wires to vfile.
 *
 * @param file - The VFile to report to
 * @param rule - The diagnostic rule definition
 * @param options - Position and context data
 */
export function report(
  file: VFile | null | undefined,
  rule: Diagnostic,
  options?: ReportOptions
): void {
  if (!file) {
    return;
  }

  // Construct the rule ID from type, namespace, and code
  const ruleId = `${rule.type}:${rule.namespace}:${rule.code}`;

  const context = options?.context ?? {};
  const message = rule.message(context);

  // Create the vfile message
  const vfileMessage = file.message(message, options?.node);

  // Set the ruleId explicitly
  vfileMessage.ruleId = ruleId;

  // Set the help URL if provided
  vfileMessage.url = rule.helpUrl;

  // Set the description
  vfileMessage.note = rule.description;

  // Set source to the namespace (middle part of ruleId)
  vfileMessage.source = rule.namespace;

  // Set the ancestors
  const baseAncestors = options?.ancestors ?? [];
  vfileMessage.ancestors = options?.node
    ? [...baseAncestors, options.node]
    : baseAncestors;

  // Set the place
  vfileMessage.place = options?.node?.position;

  // Map severity to fatal flag
  switch (rule.severity) {
    case "error":
      vfileMessage.fatal = true;
      break;
    case "warning":
      vfileMessage.fatal = false;
      break;
    case "info":
      vfileMessage.fatal = null;
      break;
    case null:
      vfileMessage.fatal = null;
      break;
    default:
      vfileMessage.fatal = undefined;
      break;
  }
}
