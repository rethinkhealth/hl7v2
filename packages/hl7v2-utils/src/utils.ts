import type { Nodes, Segment } from "@rethinkhealth/hl7v2-ast";
import type { PathParts } from "./types";

// -------------
// Delimiters
// -------------

export const DEFAULT_DELIMITERS = {
  field: "|",
  component: "^",
  repetition: "~",
  subcomponent: "&",
  escape: "\\",
  segment: "\r",
};

// -------------
// General
// -------------

/**
 * Utility: check if a node is semantically empty
 */
export function isEmptyNode(node: Nodes | null | undefined): boolean {
  if (!node) {
    return true;
  }

  // If node has a "value" property (Subcomponent, maybe Component)
  if ("value" in node) {
    return !node.value || node.value.trim() === "";
  }

  // If node has children (Field, Component, Repetition, Segment, Root, etc.)
  if ("children" in node) {
    if (!node.children || node.children.length === 0) {
      return true;
    }

    // If node has more than one child, then it is considered non-empty
    if (node.children.length > 1) {
      return false;
    }

    // If node has only one child, then it is considered empty if the child is also empty
    return isEmptyNode(node.children[0]);
  }

  // Fallback: consider unknown node as non-empty
  return false;
}

/**
 * Validates that a numeric value is a valid HL7 component number.
 *
 * @param value - The value to validate
 * @param componentName - Name of the component for error messages
 * @throws {Error} If the value is not a valid positive integer
 */
function validateHl7Number(value: number, componentName: string): void {
  if (!Number.isInteger(value)) {
    throw new Error(`${componentName} must be an integer, got: ${value}`);
  }
  if (value < 1) {
    throw new Error(
      `${componentName} must be positive (1-based), got: ${value}`
    );
  }
}

/**
 * Validates that an AST index is a valid non-negative integer.
 *
 * @param value - The value to validate
 * @param componentName - Name of the component for error messages
 * @throws {Error} If the value is not a valid non-negative integer
 */
function validateAstIndex(value: number, componentName: string): void {
  if (!Number.isInteger(value)) {
    throw new Error(`${componentName} must be an integer, got: ${value}`);
  }
  if (value < 0) {
    throw new Error(
      `${componentName} must be non-negative (0-based), got: ${value}`
    );
  }
}

// Regex pattern for valid segment IDs (2-4 uppercase letters/numbers)
const SEGMENT_ID_PATTERN = /^[A-Z0-9]{2,4}$/;

/**
 * Validates a segment ID string.
 *
 * @param segmentId - The segment ID to validate
 * @throws {Error} If the segment ID is invalid
 */
function validateSegmentId(segmentId: string): void {
  if (!segmentId || typeof segmentId !== "string") {
    throw new Error(`segmentId must be a non-empty string, got: ${segmentId}`);
  }
  if (segmentId.trim() !== segmentId) {
    throw new Error(
      `segmentId cannot have leading/trailing whitespace, got: "${segmentId}"`
    );
  }
  if (!SEGMENT_ID_PATTERN.test(segmentId)) {
    throw new Error(
      `segmentId must be 2-4 uppercase letters/numbers, got: "${segmentId}"`
    );
  }
}

/**
 * Extracts the segment ID from the segment's first field.
 *
 * Traverses the AST structure to find the segment identifier stored
 * in the first field's first repetition's first component's first
 * subcomponent value.
 *
 * @param seg - The segment AST node to extract the ID from
 * @returns The segment ID string (e.g., 'PID', 'MSH'), or null if not found
 *
 * @example
 * ```ts
 * const segmentId = getSegmentId(segmentNode);
 * if (segmentId === 'PID') {
 *   console.log('Processing patient identification segment');
 * }
 * ```
 */
export function getSegmentId(seg: Segment): string | null {
  return seg.children[0]?.children[0]?.children[0]?.children[0]?.value ?? null;
}

/**
 * Builds an HL7 path string from structured path components.
 *
 * Constructs a standardized HL7 path string using the format:
 * `SEGMENT-FIELD[REPETITION].COMPONENT.SUBCOMPONENT`
 *
 * **IMPORTANT: All numeric components use 1-based HL7 numbering.**
 * This function expects HL7-standard 1-based numbers, not 0-based AST indices.
 * Use `pathFromIndices()` if you have 0-based AST indices to convert.
 *
 * Components are added incrementally - if a field is not provided,
 * none of its child components will be included in the path.
 *
 * @param parts - Path components using HL7 1-based numbering
 * @param parts.segmentId - Segment identifier (required, 2-4 uppercase chars)
 * @param parts.field - HL7 field number (1-based, optional, must be ≥1)
 * @param parts.repetition - Repetition number (1-based, optional, must be ≥1)
 * @param parts.component - Component number (1-based, optional, must be ≥1)
 * @param parts.subcomponent - Subcomponent number (1-based, optional, must be ≥1)
 * @returns Formatted HL7 path string
 * @throws {Error} If any numeric component is not a positive integer
 *
 * @example
 * ```ts
 * // Segment only
 * formatPath({ segmentId: 'PID' });
 * // Result: "PID"
 *
 * // Field path (HL7 field 5 = 1-based)
 * formatPath({ segmentId: 'PID', field: 5 });
 * // Result: "PID-5"
 *
 * // Complete path with HL7 1-based numbering
 * formatPath({
 *   segmentId: 'PID',
 *   field: 5,        // HL7 field 5 (1-based)
 *   repetition: 1,   // First repetition (1-based)
 *   component: 2,    // Second component (1-based)
 *   subcomponent: 1  // First subcomponent (1-based)
 * });
 * // Result: "PID-5[1].2.1"
 *
 * // Error cases - these will throw:
 * formatPath({ segmentId: 'PID', field: 0 });        // Error: field must be ≥1
 * formatPath({ segmentId: 'PID', field: 1.5 });      // Error: must be integer
 * formatPath({ segmentId: '' });                     // Error: invalid segment ID
 * ```
 */
export function formatPath(parts: PathParts): string {
  const { segmentId, field, repetition, component, subcomponent } = parts;

  // Validate segment ID
  validateSegmentId(segmentId);

  // Validate numeric components (all should be 1-based positive integers)
  if (field != null) {
    validateHl7Number(field, "field");
  }
  if (repetition != null) {
    validateHl7Number(repetition, "repetition");
  }
  if (component != null) {
    validateHl7Number(component, "component");
  }
  if (subcomponent != null) {
    validateHl7Number(subcomponent, "subcomponent");
  }

  // Build path string
  let out = segmentId;
  if (typeof field === "number") {
    out += `-${field}`;
  }
  if (typeof repetition === "number") {
    out += `[${repetition}]`;
  }
  if (typeof component === "number") {
    out += `.${component}`;
  }
  if (typeof subcomponent === "number") {
    out += `.${subcomponent}`;
  }
  return out;
}

/**
 * Converts AST indices to 1-based HL7 path format, including only present components.
 *
 * Helper function for converting AST traversal indices to HL7 path strings.
 * Only includes path components that are explicitly provided (not undefined).
 * This prevents generating incomplete paths like "PID-[1]" when field is missing.
 *
 * **IMPORTANT: Mixed indexing convention:**
 * - `fieldIndex`: HL7 field number (1-based)
 * - Other indices: 0-based AST indices, converted to 1-based HL7 numbering
 *
 * @param parts - AST index-based path components
 * @param parts.segmentId - Segment identifier (required, 2-4 uppercase chars)
 * @param parts.fieldIndex - HL7 field number (1-based, optional, must be ≥1)
 * @param parts.repetitionIndex - 0-based repetition index (optional, must be ≥0)
 * @param parts.componentIndex - 0-based component index (optional, must be ≥0)
 * @param parts.subcomponentIndex - 0-based subcomponent index (optional, must be ≥0)
 * @returns Formatted HL7 path string with only present components
 * @throws {Error} If any provided index is invalid
 *
 * @example
 * ```ts
 * // Full path with all components
 * pathFromIndices({
 *   segmentId: 'PID',
 *   fieldIndex: 4,        // HL7 field number -> HL7 field 4
 *   repetitionIndex: 0,   // 0-based AST index -> HL7 [1]
 *   componentIndex: 1     // 0-based AST index -> HL7 .2
 * });
 * // Result: "PID-4[1].2"
 *
 * // Partial path - only includes present components
 * pathFromIndices({
 *   segmentId: 'PID',
 *   fieldIndex: 5         // Only field provided
 * });
 * // Result: "PID-5" (not "PID-5[undefined]")
 *
 * pathFromIndices({
 *   segmentId: 'PID'      // Only segment provided
 * });
 * // Result: "PID" (segment-level path)
 *
 * // Common usage during AST traversal
 * visit(root, (node, index, parent) => {
 *   if (isComponent(node)) {
 *     const path = pathFromIndices({
 *       segmentId: getSegmentId(segmentNode),
 *       fieldIndex: fieldNumber,      // HL7 field number (1-based)
 *       repetitionIndex: repIndex,    // 0-based from parent.children[repIndex]
 *       componentIndex: compIndex     // 0-based from parent.children[compIndex]
 *       // subcomponentIndex omitted if not relevant
 *     });
 *
 *     reportDiagnostic(file, {
 *       code: 'invalid_value',
 *       message: 'Component value is invalid',
 *       source: 'field',
 *       path // Clean path: "PID-4[1].2"
 *     }, { ruleId: 'field:value' });
 *   }
 * });
 * ```
 */
export function pathFromIndices(parts: {
  segmentId: string;
  fieldIndex?: number; // HL7 field number (1-based)
  repetitionIndex?: number; // 0-based AST index
  componentIndex?: number; // 0-based AST index
  subcomponentIndex?: number; // 0-based AST index
}): string {
  const {
    segmentId,
    fieldIndex,
    repetitionIndex,
    componentIndex,
    subcomponentIndex,
  } = parts;

  // Validate segment ID
  validateSegmentId(segmentId);

  // Validate provided indices (undefined values are skipped)
  if (fieldIndex != null) {
    validateHl7Number(fieldIndex, "fieldIndex"); // fieldIndex is HL7 field number (1-based)
  }
  if (repetitionIndex != null) {
    validateAstIndex(repetitionIndex, "repetitionIndex");
  }
  if (componentIndex != null) {
    validateAstIndex(componentIndex, "componentIndex");
  }
  if (subcomponentIndex != null) {
    validateAstIndex(subcomponentIndex, "subcomponentIndex");
  }

  // Build path with only present components
  return formatPath({
    segmentId,
    field: fieldIndex, // fieldIndex is already HL7 field number (or undefined)
    repetition: repetitionIndex != null ? repetitionIndex + 1 : undefined,
    component: componentIndex != null ? componentIndex + 1 : undefined,
    subcomponent: subcomponentIndex != null ? subcomponentIndex + 1 : undefined,
  });
}
