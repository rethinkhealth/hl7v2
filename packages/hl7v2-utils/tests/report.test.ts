import { VFile } from "vfile";
import { describe, expect, it } from "vitest";
import { report } from "../src/report";
import type { Diagnostic } from "../src/types";

describe("report", () => {
  it("calls the rule message function with context", () => {
    const file = new VFile();
    const rule: Diagnostic = {
      type: "validator",
      namespace: "order",
      code: "transition",
      title: "Invalid Segment Transition",
      description: "A segment arrived that is not allowed.",
      severity: "error",
      message: (ctx: Record<string, unknown>) =>
        `Segment '${ctx.actual}' not allowed`,
    };

    report(file, rule, {
      context: { actual: "EVN" },
    });

    expect(file.messages).toHaveLength(1);
    expect(file.messages[0]?.message).toBe("Segment 'EVN' not allowed");
  });

  it("extracts namespace as source from rule", () => {
    const file = new VFile();
    const rule: Diagnostic = {
      type: "lint",
      namespace: "field",
      code: "required",
      title: "Required Field Missing",
      description: "A required field is missing.",
      severity: "warning",
      message: () => "Field required",
    };

    report(file, rule);

    expect(file.messages[0]?.source).toBe("field");
  });

  it("sets ruleId from type:namespace:code", () => {
    const file = new VFile();
    const rule: Diagnostic = {
      type: "validator",
      namespace: "order",
      code: "acceptance",
      title: "Message Incomplete",
      description: "Message did not reach accepting state.",
      severity: "error",
      message: () => "Message incomplete",
    };

    report(file, rule);

    expect(file.messages[0]?.ruleId).toBe("validator:order:acceptance");
  });

  it("maps error severity to fatal: true", () => {
    const file = new VFile();
    const rule: Diagnostic = {
      type: "validator",
      namespace: "order",
      code: "transition",
      title: "Invalid Segment Transition",
      description: "A segment arrived that is not allowed.",
      severity: "error",
      message: () => "Error message",
    };

    report(file, rule);

    expect(file.messages[0]?.fatal).toBe(true);
  });

  it("maps warning severity to fatal: false", () => {
    const file = new VFile();
    const rule: Diagnostic = {
      type: "validator",
      namespace: "order",
      code: "unknown",
      title: "Unknown Segment",
      description: "Segment not recognized.",
      severity: "warning",
      message: () => "Warning message",
    };

    report(file, rule);

    expect(file.messages[0]?.fatal).toBe(false);
  });

  it("maps info severity to fatal: null", () => {
    const file = new VFile();
    const rule: Diagnostic = {
      type: "validator",
      namespace: "order",
      code: "info",
      title: "Info",
      description: "Informational message.",
      severity: "info",
      message: () => "Info message",
    };

    report(file, rule);

    expect(file.messages[0]?.fatal).toBeNull();
  });

  it("passes context to message function", () => {
    const file = new VFile();
    const rule: Diagnostic = {
      type: "lint",
      namespace: "field",
      code: "required",
      title: "Required Field Missing",
      description: "A required field is missing.",
      severity: "error",
      message: (ctx: Record<string, unknown>) =>
        `Field '${ctx.fieldPath}' is required. Value: ${ctx.value}`,
    };

    report(file, rule, {
      context: { fieldPath: "PID-5", value: "test" },
    });

    expect(file.messages[0]?.message).toBe(
      "Field 'PID-5' is required. Value: test"
    );
  });

  it("handles arrays in context by joining them", () => {
    const file = new VFile();
    const rule: Diagnostic = {
      type: "validator",
      namespace: "order",
      code: "transition",
      title: "Invalid Segment Transition",
      description: "A segment arrived that is not allowed.",
      severity: "error",
      message: (ctx: Record<string, unknown>) =>
        `Expected: ${Array.isArray(ctx.expected) ? (ctx.expected as string[]).join(", ") : ctx.expected}`,
    };

    report(file, rule, {
      context: { expected: ["PID", "NK1", "PV1"] },
    });

    expect(file.messages[0]?.message).toBe("Expected: PID, NK1, PV1");
  });

  it("ignores null file gracefully", () => {
    const rule: Diagnostic = {
      type: "validator",
      namespace: "order",
      code: "transition",
      title: "Invalid Segment Transition",
      description: "A segment arrived that is not allowed.",
      severity: "error",
      message: () => "Invalid transition",
    };

    // Should not throw
    expect(() => {
      report(null, rule);
    }).not.toThrow();

    expect(() => {
      report(undefined, rule);
    }).not.toThrow();
  });

  it("sets position when provided", () => {
    const file = new VFile();
    const rule: Diagnostic = {
      type: "validator",
      namespace: "order",
      code: "transition",
      title: "Invalid Segment Transition",
      description: "A segment arrived that is not allowed.",
      severity: "error",
      message: () => "Invalid transition",
      helpUrl: "https://example.com/transition",
    };
    const position = {
      start: { line: 2, column: 5 },
      end: { line: 2, column: 5 },
    };
    const node =  {
        type: "segment",
        position,
      }

    // When
    report(file, rule, {
      node,
    });

    // vfile stores position internally; just verify message was created
    expect(file.messages[0]).toMatchObject({
      message: rule.message({}),
      ancestors: [node],
      ruleId: "validator:order:transition",
      url: "https://example.com/transition",
      note: "A segment arrived that is not allowed.",
      source: "order",
      fatal: true,
    });
  });

  it("works without context", () => {
    const file = new VFile();
    const rule: Diagnostic = {
      type: "validator",
      namespace: "order",
      code: "acceptance",
      title: "Message Incomplete",
      description: "Message did not reach accepting state.",
      severity: "error",
      message: () => "Message incomplete",
    };

    expect(() => {
      report(file, rule);
    }).not.toThrow();

    expect(file.messages[0]?.message).toBe("Message incomplete");
  });
});
