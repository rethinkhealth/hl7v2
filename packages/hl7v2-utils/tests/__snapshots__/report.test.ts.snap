// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`report > matches the vfile message snapshot 1`] = `
"[
  {
    "ancestors": [
      {
        "type": "segment",
        "position": {
          "start": {
            "line": 2,
            "column": 5
          },
          "end": {
            "line": 2,
            "column": 25
          }
        }
      }
    ],
    "column": 5,
    "fatal": true,
    "file": "",
    "message": "Expected PID, NK1, got 'EVN'",
    "line": 2,
    "name": "2:5-2:25",
    "place": {
      "start": {
        "line": 2,
        "column": 5
      },
      "end": {
        "line": 2,
        "column": 25
      }
    },
    "reason": "Expected PID, NK1, got 'EVN'",
    "ruleId": "validator:order:transition",
    "source": "order",
    "note": "A segment arrived that is not allowed.",
    "url": "https://example.com/transition"
  }
]"
`;
